kratos/solving_strategies/builder_and_solvers/p_multigrid/diagonal_scaling.cpp

```diff
@@ -0,0 +1,174 @@
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Máté Kelemen
+//
+
+// External includes
+#include "tinyexpr/tinyexpr.h" // te_expr, te_compile, te_eval, te_free
+
+// Project includes
+#include "solving_strategies/builder_and_solvers/p_multigrid/diagonal_scaling.hpp" // DiagonalScaling, ParseDiagonalScaling, GetDiagonalScaleFactor
+#include "spaces/ublas_space.h" // TUblasSparseSpace
+#include "utilities/reduction_utilities.h" // MaxReduction, AbsMaxReduction
+#include "utilities/profiler.h" // KRATOS_PROFILE_SCOPE
+
+// STL includes
+#include <optional> // std::optional
+
+
+namespace Kratos {
+
+
+struct Scaling::Impl {
+    using ExpressionPtr = std::unique_ptr<te_expr,std::function<void(te_expr*)>>;
+
+    std::optional<double> mMaybeAbsMax;
+
+    std::optional<double> mMaybeNorm;
+
+    mutable ExpressionPtr mpExpression;
+}; // struct Scaling::Impl
+
+
+Scaling::Scaling()
+    : Scaling(Parameters())
+{
+}
+
+
+Scaling::Scaling(Parameters Settings)
+    : mpImpl(new Impl {/*mMaybeAbsMax=*/0.0,
+                       /*mMaybeNorm=*/  0.0,
+                       /*mpExpression*/ Impl::ExpressionPtr(nullptr, [](te_expr*){})})
+{
+    // Sanity checks.
+    KRATOS_ERROR_IF_NOT(Settings.IsNumber() || Settings.Is<std::string>())
+        << "Expecting a numeric literal or a string, but got " << Settings;
+
+    KRATOS_TRY
+
+    std::string expression;
+    if (Settings.IsNumber()){
+        expression = std::to_string(Settings.Get<double>());
+    } /*if Settings.IsNumber*/ else {
+        expression = Settings.Get<std::string>();
+    } /*if Settings.Is<std::string>*/
+
+    te_variable variables[] = {
+        {"max",  &mpImpl->mMaybeAbsMax.value()},
+        {"norm", &mpImpl->mMaybeNorm.value()}
+    };
+
+    int error = 0;
+    mpImpl->mpExpression = Impl::ExpressionPtr(/*function ptr: */te_compile(expression.c_str(), variables, 2, &error),
+                                               /*deleter: */     [](te_expr* p){if (p) te_free(p);});
+    KRATOS_ERROR_IF_NOT(mpImpl->mpExpression.get())
+        << "Failed to compile expression (error at position " << error - 1 << " in \"" << expression << "\"). "
+        << "Make sure the expression is compatible with TinyExpr and is a function of "
+        << "only \"max\" and \"norm\".";
+
+    this->ClearCache();
+    KRATOS_CATCH("")
+}
+
+
+Scaling::~Scaling() = default;
+
+
+void Scaling::ClearCache()
+{
+    mpImpl->mMaybeAbsMax.reset();
+    mpImpl->mMaybeNorm.reset();
+}
+
+
+double Scaling::Evaluate() const
+{
+    KRATOS_ERROR_IF_NOT(mpImpl->mMaybeAbsMax.has_value() && mpImpl->mMaybeNorm.has_value())
+        << "Matrix diagonal norms are not cached. Make sure to invoke Scaling::Cache before Scaling::Evaluate.";
+
+    return te_eval(mpImpl->mpExpression.get());
+}
+
+
+void Scaling::Cache(double DiagonalMax, double DiagonalNorm)
```

Comment:
I would use `SetPossibleNorms` or something similar rather than `Cache`.

Reply:
I'm okay with changing the name but then it has to be more descriptive of what's happening.

The behavior (and related dangers) is pretty well communicated by "caching" in my opinion. It makes it crystal clear that these values are never computed on the fly, but pulled in from stored data. `SetPossibleNorms` on the other hand sounds like it's setting the **method** by which the scaling should be computed and completely misses the caching behavior.

Any suggestions?